### PR TITLE
[RLlib] Fix accumulation of results in (new API stack) Algorithm.

### PR DESCRIPTION
**Pull request by sven1977**
1) Change `Algorithm.training_step()` API to NOT return anything anymore on the new API stack (old stack remains as-is).
2) Instead, all results should be logged into the Algorithm's `MetricsLogger` instance (self.metrics).
3) The Algorithm now takes care of properly accumulating `training_step` stats for iterations, which contain more than 1 call to `training_step`. This was a bug before. It didn't matter so much for loss stats (which should be averaged over the n training_step calls but were reported as only the last call's results), but makes a big difference for stats that have reduce="sum" (counters), where the counts for earlier `training_step` calls were completely ignored.
4) EnvRunners keep track of their own lifetime counts (e.g. NUM_ENV_STEPS_SAMPLED_LIFETIME). These counts are no longer synched /broadcast back from the algo each iteration. This way, the algo carries the global (across all env runners) `NUM_ENV_STEPS_SAMPLED_LIFETIME` count and each env runner only carries its own lifetime `NUM_ENV_STEPS_SAMPLED_LIFETIME` count.
5) The root-level `NUM_ENV_STEPS_SAMPLED_LIFETIME` result key (returned by `algo.step()`) is a mere duplicate now of `ENV_RUNNER_RESULTS/NUM_ENV_STEPS_SAMPLED_LIFETIME` and should no longer be used (will be deprecated soon).
6) Bug in `ENV_RUNNER_RESULTS/NUM_ENV_STEPS_SAMPLED_LIFETIME` (quadratic/exponential growth) has been fixed.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/single_agent_env_runner.py [hunk 1]
```diff
@@ -494,12 +497,14 @@ def _sample_episodes(
 
                 # RLModule forward pass: Explore or not.
                 if explore:
-                    env_steps_lifetime = (
+                    # Global env steps sampled are (roughly) this EnvRunner's lifetime
```
Comment (sven1977) [inline rllib/env/single_agent_env_runner.py:500]: EnvRunners do NOT carry the global timesteps anymore. Instead, they only keep track of their very own lifetime counts and estimate the global lifetime steps through simply multiplying with the number of EnvRunners in the algo. This way, we can simplify and separate concerns even more and a) skip the extra broadcast of global timesteps from algo back to all workers and b) more naturally "reduce" the lifetime steps on the algo side by simply adding over the individual EnvRunners' results.

File: rllib/algorithms/ppo/ppo.py [hunk 2]
```diff
@@ -435,53 +425,29 @@ def _training_step_new_api_stack(self) -> ResultDict:
...
             # Return early if all our workers failed.
             if not episodes:
-                return {}
+                return
 
             # Reduce EnvRunner metrics over the n EnvRunners.
             self.metrics.merge_and_log_n_dicts(
                 env_runner_results, key=ENV_RUNNER_RESULTS
             )
-            # Log lifetime counts for env- and agent steps.
```
Comment (sven1977) [inline rllib/algorithms/ppo/ppo.py:444]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/marwil/marwil.py [hunk 3]
```diff
@@ -467,22 +460,7 @@ class (multi-/single-learner setup) and evaluation on
 
             # Log training results.
             self.metrics.merge_and_log_n_dicts(learner_results, key=LEARNER_RESULTS)
-            self.metrics.log_value(
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:470]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/impala/impala.py [hunk 4]
```diff
@@ -647,23 +644,6 @@ def training_step(self) -> ResultDict:
                 ),
             )
 
-        # Log lifetime counts for env- and agent steps.
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:650]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 5]
```diff
@@ -581,23 +582,6 @@ def training_step(self) -> ResultDict:
                     self.replay_buffer.add(episodes=_episodes)
                     total_sampled += sum(len(eps) for eps in _episodes)
 
-                # Update lifetime counts (now that we gathered results from all
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/dreamerv3.py:584]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/dqn/dqn.py [hunk 6]
```diff
@@ -775,29 +752,6 @@ def _training_step_new_api_stack(self, *, with_noise_reset) -> ResultDict:
                     self.metrics.merge_and_log_n_dicts(
                         learner_results, key=LEARNER_RESULTS
                     )
-                    self.metrics.log_value(
```
Comment (sven1977) [inline rllib/algorithms/dqn/dqn.py:778]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/dqn/dqn.py [hunk 7]
```diff
@@ -688,38 +680,16 @@ def _training_step_new_api_stack(self, *, with_noise_reset) -> ResultDict:
             with self.metrics.log_time((TIMERS, REPLAY_BUFFER_ADD_DATA_TIMER)):
                 self.local_replay_buffer.add(episodes)
 
-        self.metrics.log_dict(
```
Comment (sven1977) [inline rllib/algorithms/dqn/dqn.py:691]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/cql/cql.py [hunk 8]
```diff
@@ -340,22 +334,6 @@ def _training_step_new_api_stack(self) -> ResultDict:
 
             # Log training results.
             self.metrics.merge_and_log_n_dicts(learner_results, key=LEARNER_RESULTS)
-            self.metrics.log_value(
```
Comment (sven1977) [inline rllib/algorithms/cql/cql.py:343]: these bulky blocks inside each algo's training_step() are no longer necessary

File: rllib/algorithms/appo/appo.py [hunk 9]
```diff
@@ -288,60 +285,51 @@ def __init__(self, config, *args, **kwargs):
...
-        train_results = super().training_step()
+    def training_step(self) -> None:
+        if self.config.enable_rl_module_and_learner:
+            return super().training_step()
 
+        train_results = super().training_step()
         # Update the target network and the KL coefficient for the APPO-loss.
         # The target network update frequency is calculated automatically by the product
         # of `num_epochs` setting (usually 1 for APPO) and `minibatch_buffer_size`.
-        if self.config.enable_rl_module_and_learner:
```
Comment (sven1977) [inline rllib/algorithms/appo/appo.py:297]: no more hybrid stack

File: rllib/algorithms/algorithm.py [hunk 10]
```diff
@@ -826,7 +826,7 @@ def setup(self, config: AlgorithmConfig) -> None:
             self.env_runner_group.sync_env_runner_states(
                 config=self.config,
                 env_steps_sampled=self.metrics.peek(
-                    NUM_ENV_STEPS_SAMPLED_LIFETIME, default=0
+                    (ENV_RUNNER_RESULTS, NUM_ENV_STEPS_SAMPLED_LIFETIME), default=0
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:829]: So, we do still sync the metrics across `EnvRunner`s?

File: rllib/algorithms/algorithm.py [hunk 11]
```diff
@@ -1662,7 +1663,7 @@ def restore_workers(self, workers: EnvRunnerGroup) -> None:
             )
 
     @OverrideToImplementCustomLogic
-    def training_step(self) -> ResultDict:
+    def training_step(self) -> None:
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:1666]: `Union[ResultDict, None]`?

File: rllib/algorithms/algorithm.py [hunk 12]
```diff
@@ -3188,633 +3189,703 @@ def validate_env(env: EnvType, env_context: EnvContext) -> None:
...
-    def __getstate__(self) -> Dict:
-        """Returns current state of Algorithm, sufficient to restore it from scratch.
+        Calls `self.training_step()` repeatedly until the configured minimum time (sec),
+        minimum sample- or minimum training steps have been reached.
 
         Returns:
-            The current state dict of this Algorithm, which can be used to sufficiently
-            restore the algorithm from scratch without any other information.
+            The ResultDict from the last call to `training_step()`. Note that even
+            though we only return the last ResultDict, the user stil has full control
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:3200]: Small  nit: "stil" -> "still"

File: rllib/algorithms/algorithm.py [hunk 13]
```diff
@@ -3188,633 +3189,703 @@ def validate_env(env: EnvType, env_context: EnvContext) -> None:
...
+        eval_results[
+            "num_in_flight_async_reqs"
+        ] = self.eval_env_runner_group.num_in_flight_async_reqs()
+        eval_results[
+            "num_remote_worker_restarts"
+        ] = self.eval_env_runner_group.num_remote_worker_restarts()
+
+        return train_results, {"evaluation": eval_results}, train_iter_ctx
+
+    def _run_offline_evaluation(self):
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:3447]: Oh, we have to implement this one for the new stack

File: rllib/algorithms/algorithm.py [hunk 14]
```diff
@@ -3188,633 +3189,703 @@ def validate_env(env: EnvType, env_context: EnvContext) -> None:
...
+        # Resolve all `Stats` leafs by peeking (get their reduced values).
+        return tree.map_structure(
+            lambda s: s.peek() if isinstance(s, Stats) else s,
+            results,
+        )
 
-        if state["config"].enable_rl_module_and_learner:
-            state["learner_state_dir"] = os.path.join(
-                checkpoint_info["checkpoint_dir"], "learner"
+    def __repr__(self):
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:3532]: Very nice!

File: rllib/algorithms/algorithm.py [hunk 15]
```diff
@@ -3188,633 +3189,703 @@ def validate_env(env: EnvType, env_context: EnvContext) -> None:
...
-        env_steps = agent_steps = 0
+        msgpack = None
+        if checkpoint_info.get("format") == "msgpack":
+            msgpack = try_import_msgpack(error=True)
 
-        # Kick off sampling on all evaluation workers (async).
-        # How many timesteps do we need to run?
-        eval_cfg = self.evaluation_config
-        time_out = self.config.evaluation_sample_timeout_s
+        with open(checkpoint_info["state_file"], "rb") as f:
```
Comment (simonsays1980) [inline rllib/algorithms/algorithm.py:3742]: Can we check, if cloud checkpointing still works?

File: rllib/algorithms/cql/cql.py [hunk 16]
```diff
@@ -340,22 +334,6 @@ def _training_step_new_api_stack(self) -> ResultDict:
...
-            )
-            self.metrics.log_dict(
-                {
-                    (LEARNER_RESULTS, mid, NUM_MODULE_STEPS_TRAINED_LIFETIME): (
-                        stats[NUM_MODULE_STEPS_TRAINED]
-                    )
-                    for mid, stats in self.metrics.peek(LEARNER_RESULTS).items()
-                },
-                reduce="sum",
-            )
```
Comment (simonsays1980) [inline rllib/algorithms/cql/cql.py:358]: Awesome change!

File: rllib/algorithms/marwil/marwil.py [hunk 17]
```diff
@@ -428,21 +424,18 @@ def get_default_policy_class(
...
-            return self._training_step_old_api_stack()
-
-    def _training_step_new_api_stack(self) -> ResultDict:
+    def training_step(self) -> None:
         """Implements training logic for the new stack
 
         Note, this includes so far training with the `OfflineData`
         class (multi-/single-learner setup) and evaluation on
         `EnvRunner`s. Note further, evaluation on the dataset itself
         using estimators is not implemented, yet.
```
Comment (simonsays1980) [inline rllib/algorithms/marwil/marwil.py:433]: Exactly. Another item on the list!

File: rllib/env/single_agent_env_runner.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/env/single_agent_env_runner.py:500]: @sven1977 This means that epsilon greedy will now depend on a single runner's env steps sampled over its lifetime and decay slower. Did we already test in DQN what influence this has?

File: rllib/utils/metrics/metrics_logger.py [hunk 18]
```diff
@@ -34,16 +34,71 @@ class MetricsLogger:
...
-        logger.log_value("loss", 0.001)
-        logger.log_value("loss", 0.002)
-        logger.log_value("loss", 0.003)
+        # 1) Logging float values (mean over window):
+        # Log some loss under the "loss" key. By default, all logged values
+        # under that key are averaged and reported back, once `reduce()` is called.
+        logger.log_value("loss", 0.001, reduce="mean", window=10)
+        logger.log_value("loss", 0.002)  # <- no need to repeat arg/options on same key
         # Peek at the current (reduced) value of "loss":
+        check(logger.peek("loss"), 0.0015)  # <- expect average value
```
Comment (simonsays1980) [inline rllib/utils/metrics/metrics_logger.py:49]: We should add to the docs sth like an FAQ for MetricsLogger. Like "I see NaN values in my Evaluation results for each non-evaluated Iteration. What shall I do?"

**Review by simonsays1980 — APPROVED** [review]
LGTM. Some questions remain and we should test, if cloud checkpointing still works with this. Also epsilon greedy will behave differently now, I guess.

File: rllib/env/single_agent_env_runner.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/single_agent_env_runner.py:500]: I think it's fine. The EnvRunners are now doing this here:
```
global_env_steps_lifetime = (
                        self.metrics.peek(NUM_ENV_STEPS_SAMPLED_LIFETIME, default=0)
                        + ts
                    ) * (self.config.num_env_runners or 1)  <- !!HERE!! (multiply by number of env runners to get a rough estimate of the overall timesteps of all env-runners
```

File: rllib/algorithms/algorithm.py (hunk 10 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:829]: We do, but only in `setup()` (at very beginning).

Actually, good point. Let me check, whether we properly sync these stats after a worker restart and/or after a checkpoint restore ...

File: rllib/algorithms/algorithm.py (hunk 11 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:1666]: You are right, but let's already go all-in on the new API stack and ignore the old APIs.

File: rllib/algorithms/algorithm.py (hunk 12 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:3200]: fixed :)

File: rllib/algorithms/algorithm.py (hunk 13 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:3447]: yup, I guess. If we want to support these evaluation algos. I'm not sure yet how much of a priority this should be, though ...

File: rllib/algorithms/algorithm.py (hunk 15 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:3742]: Hmm, I just moved around some methods to organize better (old API stack methods to the bottom). I didn't touch that functionality.
